enarx-keep-sev-shim/src/syscall.rs

```diff
@@ -85,6 +90,35 @@ impl Handler {
         hostcall::shim_exit(self.a as _);
     }
 
+    pub fn read(&self) -> Result<usize, libc::c_int> {
+        let fd = self.a;
+        let data = self.b as *mut u8;
+        let len = self.c;
+        let mut host_call = HOST_CALL.try_lock().ok_or(libc::EIO)?;
+
+        let trusted: &mut [u8] = unsafe { core::slice::from_raw_parts_mut(data, len) };
+
+        let block = host_call.as_mut_block();
+
+        let c = block.cursor();
+        let (_, buf) = unsafe { c.alloc::<u8>(trusted.len()).or(Err(libc::EMSGSIZE))? };
+
+        let buf_address = Address::from(buf.as_ptr());
+        let shim_virt_address = ShimVirtAddr::try_from(buf_address).map_err(|_| libc::EFAULT)?;
+        let host_virt: HostVirtAddr<_> = ShimPhysAddr::from(shim_virt_address).into();
+
+        block.msg.req = request!(libc::SYS_read => fd, host_virt, len);
+        let result = unsafe { host_call.hostcall() };
+        let result = result.map(|r| r[0].into())?;
+
```

Comment:
Before copying from the untrusted sallyport buffer to trusted, should it check that the number of bytes read is the expected number and does not exceed the trusted buffer size? In SGX we have something like
```
if trusted.len() < ret[0].into() {
            self.attacked();
}
```

Reply:
right... fixed :+1: